internal/project/project.go

```diff
@@ -372,14 +373,23 @@ func (p *Project) getScriptKind(fileName string) core.ScriptKind {
 }
 
 func (p *Project) markFileAsDirty(path tspath.Path) {
-	p.markAsDirty()
+	p.dirtyStateMu.Lock()
+	defer p.dirtyStateMu.Unlock()
+	if !p.dirty {
+		p.dirty = true
+		p.dirtyFilePath = path
```

Comment:
I think this may need to be a `Set` or something; file watching can cause this to be overwritten with subsequent updates, or even open files on branch switches or similar.

Reply:
Right, but the minute *anything* else changes, we clear out `p.dirtyFilePath` and `updateGraph` will then completely re-create the program.

Reply:
Ah, I see what you mean in `updateProgram`; you're right that it's safe. I guess we don't need to be super optimal (though I think it might just be a couple lines to keep track of multiple dirty files).

Reply:
Is this temporary? because we have this https://github.com/microsoft/TypeScript/blob/main/src/compiler/program.ts#L2458 in strada thats much more resilient to multiple file changes (can happen as part of open and close when navigating source code)